agent/app/agent_unix.go

```diff
@@ -0,0 +1,57 @@
+// +build linux
+
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package app
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/cihub/seelog"
+)
+
+// control is here to help create the cgroup root '/ecs' when
+// the feature is enabled. It also aids in unit testing
+var control cgroup.Control
+
+// initPlatformResources initializes platform resource controllers
+// Currently it helps initialize cgroup
+func (agent *ecsAgent) initPlatformResources() {
+	if agent.cfg.TaskCPUMemLimit {
+		setControl(cgroup.New())
+	}
+}
+
+// setupPlatformResources helps set up the platform resource controllers
+// Currently used to setup '/ecs' cgroup
+func (agent *ecsAgent) setupPlatformResources() error {
+	if agent.cfg.TaskCPUMemLimit {
+		return setupTaskCgroupPrefix()
+	}
+	return nil
+}
+
+// setupTaskCgroupPrefix creates the '/ecs' cgroup
+func setupTaskCgroupPrefix() error {
+	if control.Exists(config.DefaultTaskCgroupPrefix) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", config.DefaultTaskCgroupPrefix)
+		return nil
+	}
+	return control.Init()
+}
+
+// setControl sets the cgroup controller
+func setControl(controller cgroup.Control) {
```

Comment:
I'm really really really apprehensive about this pattern:
1. Using global variables even if it aids unit testing seems like an anti-pattern to me
2. This leads to all kinds of (potential) race conditions without being protected with a lock. 

Can you instead make `control` a member of `ecsAgent` and pass it around instead?

Reply:
`ecsAgent` is platform agnostic while `control` here is linux only. 
Embedding the `control` object within another was called out as an overkill.

Reply:
Would an interface help you here?

Reply:
I'm not sure if another interface is going to add significant value here. 
Also, if I eliminate the `init` items here, testing `doStart()` is going to be harder. 

I'm totally not in favor of using global objects. It just simplifies things here.
Could you please comment on what races you foresee with this approach ?

Reply:
> I'm not sure if another interface is going to add significant value here.
It will help you with not having to use this global variable

> Also, if I eliminate the `init` items here, testing `doStart()` is going to be harder. 
That should be fine. 

> Could you please comment on what races you foresee with this approach ?
As per my earlier comment, the access to `control` variable is not protected with a lock. Someone in the future will write a test where this inadvertently  gets invoked at the same time when `setupTaskCgroupPrefix` gets invoked.

Reply:
`control` is already a `linux only` interface as it does not apply to `windows`.

Reply:
So, you can have a dummy object in `ecsAgent`, which can remain uninitialized in windows.

Reply:
This is exactly the kind of problem we wanted to avoid in the first place. 

If we plan to embed this object into another or create dummies for unsupported platforms, we could have gone with a `platformResources` object.